lib/systems/default.nix

```
@@ -1,9 +1,100 @@
+ selectEmulator = platform: pkgs:
+   let
+     qemu-user = pkgs.qemu.override {
+       smartcardSupport = false;
+       spiceSupport = false;
+       openGLSupport = false;
+       virglSupport = false;
+       vncSupport = false;
+       gtkSupport = false;
+       sdlSupport = false;
+       pulseSupport = false;
+       pipewireSupport = false;
+       smbdSupport = false;
+       seccompSupport = false;
+       enableDocs = false;
+       hostCpuTargets = [ "${platform.qemuArch}-linux-user" ];
+     };
+     wine = (pkgs.winePackagesFor "wine${toString platform.parsed.cpu.bits}").minimal;
+   in
+     if lib.systems.canExecute pkgs.stdenv.hostPlatform platform
+     then "${pkgs.runtimeShell} -c '\"$@\"' --"
+     else if platform.isWindows
+     then "${wine}/bin/wine${lib.optionalString (platform.parsed.cpu.bits == 64) "64"}"
+     else if platform.isLinux && pkgs.stdenv.hostPlatform.isLinux && platform.qemuArch != null
+     then "${qemu-user}/bin/qemu-${platform.qemuArch}"
+     else if platform.isWasi
+     then "${pkgs.wasmtime}/bin/wasmtime"
+     else if platform.isMmix
+     then "${pkgs.mmixware}/bin/mmix"
+     else null;
 
-rec {
   doubles = import ./doubles.nix { inherit lib; };
+
   parse = import ./parse.nix { inherit lib; };
+
+  isCompatiblePlaceholder      = _: throw "2022-05-23: isCompatible has been removed in favor of canExecute, refer to the 22.11 changelog for details";
+
+  # These definitions have to be floated out into a `let … in` binding to ensure
+  # that they are the same exact value struct in the evaluating Nix
+  # implementation. Due to a quirk in how equality is implemented, values including
+  # functions are considered equal if they have the same memory location as long as
+  # they are inside a container value and not compared directly. Thus any attribute
+  # set will compare equal to itself even if it contains functions.
+  # By floating the functions out of the attribute set construction we'll ensure
+  # that the functions have the same memory location regardless of what call to
+  # lib.systems.elaborate produced them. This ensures that platform sets will be
+  # equal even if they have been constructed by different calls to lib.systems.elaborate
+  # —as long as their other contents are the same.
+  #
+  # For details see:
+  # - https://github.com/NixOS/nix/issues/3371
+  # - https://code.tvl.fyi/about/tvix/docs/value-pointer-equality.md
+
+  # uncomment after 24.05 branch-off
+  /*
+  canExecutePlaceholder        = _: throw "2023-06-17: `platform.canExecute` has been removed in favor of `lib.systems.canExecute platform`";
+  emulatorAvailablePlaceholder = _: throw "2023-06-17: `platform.emulatorAvailable` has been removed in favor of `lib.systems.emulatorAvailable platform`";
+  emulatorPlaceholder          = _: throw "2023-06-17: `platform.emulator` has been removed in favor of `lib.systems.emulator platform`";
+  */
+
+  # uncomment after 23.11 branch-off
+  # delete after 24.05 branch-off
```

```suggestion
  # delete canExecutePlaceholder, emulatorAvailablePlaceholder and emulatorPlaceholder after 24.05 branch-off (throwing would have been nice, but throws off `==` equality when comparing platforms elaborated on distinct but compatible Nixpkgs source trees)
```

- Won't be grouped by the comment syntax anymore, so be explicit
- Explain why not throw in this case